### PR TITLE
Added support for directly call macros defined in the same template

### LINKED ISSUE #861: Consider auto-importing macros in the same file
I've always found it a bit weird that you cannot directly call macros defined in the same file.

Having to "import" something that is in the technically in the same file feels a bit weird (especially as you import on a per-file basis).

Can't we simply auto-import them? Obviously, if you define them in the same file, you probably want to use them there.

### PR DESCRIPTION
**Pull request by hason**
Fixes #861 and #1755

### CODE REVIEW

**Comment by stof** [comment]
the issue with this is that ``submit(foo)`` is not ambiguous: if there is no function with this name, it will call the macro in the same file. but if an extension registers this function, the template will magically break because the function will be used.
and this is not consistent with the way macros imported with ``{% from ... import ... %}`` work: these ones are winning over functions (making template behavior known as you can see from the template source that this refers to an imported macro rather than a function)

**Comment by hason** [comment]
@stof good point, but similar problem is with two extensions which register the same functions.

**Comment by stof** [comment]
@hason no. In this case the behavior of templates is consistent: it always calls the function. Template authors don't care where the function is defined. The fact that an extension can overwrite previous ones is a matter of the developer, not of the template author.
With your change, a template author will create a ``submit`` macro, and then use ``submit(foo)`` as they are used to be able to use macros of the template directly. But then, they will need to figure out why the template does a totally different behavior than what the author reads from the template (it is even possible that he does not get any error if the call is compatible with the function signature). Or even worse, the template could work and then break later when a dev creates a new function (again in a way which might not generate any error message depending on signatures).

**Comment by hason** [comment]
@stof Well, what do you suggest? Should be a macro prefered?

**Comment by stof** [comment]
@hason well, this would be consistent (and we could achieve it by registering them as imported symbols), but it would be a BC break.

**Comment by fabpot** [comment]
This has been proposed before and rejected because of BC implications. Also, I very much prefer the explicit way for macros.

**Comment by hason** [comment]
@fabpot This was rejected due to named arguments for macros. This PR solves only auto-importing macros.

File: lib/Twig/ExpressionParser.php [hunk 1]
```diff
@@ -347,10 +347,25 @@ public function getFunctionNode($name, $line)
...
+                    if (!$this->parser->hasMacro($name)) {
+                        throw $e;
+                    }
+
+                    $arguments = new Twig_Node_Expression_Array(array(), $line);
+                    foreach ($this->parseArguments() as $n) {
+                        $arguments->addElement($n);
+                    }
+
+                    $node = new Twig_Node_Expression_MethodCall(new Twig_Node_Expression_Name('_self', $line), 'get'.$name, $arguments, $line);
```
Comment (stof) [inline lib/Twig/ExpressionParser.php]: this will need a change because of #1766

**Comment by fabpot** [comment]
implemented without BC breaks in #3019.

**Pull request closed without merging**